factor/token/classes/factor.php

```diff
@@ -0,0 +1,251 @@
+        return \tool_mfa\plugininfo\factor::STATE_NEUTRAL;
+    }
+
+    /**
+     * Token Implementation.
+     * We can't get_state like the parent here or it will recurse forever.
+     *
+     * @param string $state
+     * @return void
+     */
+    public function set_state($state) {
+        global $SESSION;
+        $property = 'factor_'.$this->name;
+        $SESSION->$property = $state;
+        return true;
+    }
+
+    /**
+     * Token implementation.
+     *
+     * {@inheritDoc}
+     */
+    public function possible_states($user) {
+        return [
+            \tool_mfa\plugininfo\factor::STATE_PASS,
+            \tool_mfa\plugininfo\factor::STATE_NEUTRAL,
+            \tool_mfa\plugininfo\factor::STATE_LOCKED
+        ];
+    }
+
+    /**
+     * Token implementation.
+     * Inject a checkbox into every auth form if needed.
+     *
+     * {@inheritDoc}
+     */
+    public function global_definition_after_data($mform) {
+        global $SESSION;
+
+        // First thing, we need to decide on whether we should show the checkbox.
+        $noproperty = !property_exists($SESSION, 'tool_mfa_factor_token');
+        $nostate = $this->get_state() !== \tool_mfa\plugininfo\factor::STATE_PASS;
+
+        if ($noproperty && $nostate) {
+            $expiry = get_config('factor_token', 'expiry');
+            $expirystring = format_time($expiry);
+            $mform->addElement('advcheckbox', 'factor_token_trust', '', get_string('form:trust', 'factor_token', $expirystring));
+            $mform->setType('factor_token_trust', PARAM_BOOL);
+            $mform->setDefault('factor_token_trust', true);
+        }
+    }
+
+    /**
+     * Token implementation.
+     * Store information about the token status.
+     *
+     * {@inheritDoc}
+     */
+    public function global_submit($data) {
+        global $SESSION;
+
+        // Store any kind of response here, we shouldnt show again.
+        $trust = $data->factor_token_trust;
+        $SESSION->tool_mfa_factor_token = $trust;
+    }
+
+    /**
+     * Token implementation.
+     * Pass hook to set the cookie for use in subsequent auths.
+     *
+     * {@inheritDoc}
+     */
+    public function post_pass_state() {
+        global $SESSION, $USER;
+
+        if (!property_exists($SESSION, 'tool_mfa_factor_token')) {
+            return;
+        }
+        $settoken = $SESSION->tool_mfa_factor_token;
+        if (!$settoken) {
+            return;
+        }
+
+        $userid = $USER->id;
+        $cookie = 'MFA_TOKEN_' . $userid;
+
+        list($expirytime, $expiry) = $this->calculate_expiry_time();
+
+        // Store this secret in the database.
+        $secretmanager = new secret_manager($this->name);
+        $secret = base64_encode(random_bytes(256));
+        $secretmanager->create_secret($expiry, false, $secret);
+
+        // All the prep is now done, we can set this cookie.
+        setcookie($cookie, $secret, $expirytime, '/admin/tool/mfa', '', false, true);
```

Comment:
Correct me if I'm wrong but won't this just store and verify the MFA expiry time in a cookie? i.e. If someone were to provide a valid MFA auth on their account, and edit their cookie name to pass this condition for another userid?

I don't see the userid being used as part of the verification/encrypted payload so I'm curious. That said I'm only saying what I'm seeing from the code.

Reply:
the Secret manager has implicit user checks. the userid on the token is just for deduplication of tokens if a user controls multiple lms accounts on the same browser